<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>periodicDNA • periodicDNA</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/sandstone/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="periodicDNA">
<meta property="og:description" content="periodicDNA">
<meta property="og:image" content="/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">periodicDNA</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.99.17</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/periodicDNA.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/js2264/periodicDNA/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>periodicDNA</h1>
                        <h4 class="author">Jacques Serizay</h4>
            
            <h4 class="date">2020-09-08</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/js2264/periodicDNA/blob/master/vignettes/periodicDNA.Rmd"><code>vignettes/periodicDNA.Rmd</code></a></small>
      <div class="hidden name"><code>periodicDNA.Rmd</code></div>

    </div>

    
    
<div id="introduction-to-periodicdna" class="section level3">
<h3 class="hasAnchor">
<a href="#introduction-to-periodicdna" class="anchor"></a>Introduction to periodicDNA</h3>
<p>Short DNA sequence motifs provide key information for interpreting the instructions in DNA, for example by providing binding sites for proteins or altering the structure of the double-helix. A less studied but important feature of DNA sequence motifs is their periodicity. A famous example is the 10-bp periodicity of many k-mers in nucleosome positioning (reviewed in Travers et al. 2010 or in Struhl and Segal 2013).</p>
<p>periodicDNA provides a framework to quantify the periodicity of k-mers of interest in DNA sequences. For a chosen k-mer, periodicDNA can identify which periods are statistically enriched in a set of sequences, by using a randomized shuffling approach to compute an empirical p-value. It can also generate continuous linear tracks of k-mer periodicity strength over genomic loci.</p>
</div>
<div id="internal-steps-of-periodicdna" class="section level3">
<h3 class="hasAnchor">
<a href="#internal-steps-of-periodicdna" class="anchor"></a>Internal steps of periodicDNA</h3>
<p>To estimate the periodicity strength of a given k-mer in one or several sequences, periodicDNA performs the following steps:</p>
<ol style="list-style-type: decimal">
<li>The k-mer occurrences are mapped and their pairwise distances are calculated.</li>
<li>The distribution of all the resulting pairwise distances (also called “distogram”) is generated.</li>
<li>The distogram is transformed into a frequency histogram and smoothed using a moving window of 3 to mask the universal three-base genomic periodicity. To normalize the frequency for distance decay, the local average (obtained by averaging the frequency with a moving window of 10) is then subtracted from the smoothed frequency.</li>
<li>Finally, the power spectral density (PSD) is estimated by applying a Fast Fourier Transform (Figure 1F) over the normalized frequency histogram. The magnitude of the PSD values indicates the contribution of a given period to the overall periodicity of the k-mer of interest.</li>
</ol>
<p><img src="https://raw.githubusercontent.com/js2264/periodicDNA/master/man/figures/periodicityDNA_principle.png" width="100%"></p>
</div>
<div id="quantifying-k-mer-periodicity-over-a-set-of-sequences" class="section level3">
<h3 class="hasAnchor">
<a href="#quantifying-k-mer-periodicity-over-a-set-of-sequences" class="anchor"></a>Quantifying k-mer periodicity over a set of sequences</h3>
<div id="basic-usage" class="section level4">
<h4 class="hasAnchor">
<a href="#basic-usage" class="anchor"></a>Basic usage</h4>
<p>The main goal of periodicDNA is to quantify periodicity of a given k-mer in a set of sequences. For instance, one can assess the periodicity of TT dinucleotides in sequences around TSSs of ubiquitous promoters using <code><a href="../reference/getPeriodicity.html">getPeriodicity()</a></code>.</p>
<p>In the following example, <code><a href="../reference/getPeriodicity.html">getPeriodicity()</a></code> is directly ran using a GRanges object, specifying from which genome this GRanges comes from.</p>
<div class="sourceCode" id="cb1"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="kw"><a href="http://ggplot2.tidyverse.org">ggplot2</a></span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="kw">magrittr</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="kw"><a href="https://github.com/js2264/periodicDNA">periodicDNA</a></span>)
<span class="co">#</span>
<span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span>(<span class="kw">ce11_TSSs</span>)
<span class="kw">periodicity_result</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getPeriodicity.html">getPeriodicity</a></span>(
    <span class="kw">ce11_TSSs</span>[[<span class="st">'Ubiq.'</span>]][<span class="fl">1</span><span class="op">:</span><span class="fl">500</span>],
    genome = <span class="st">'BSgenome.Celegans.UCSC.ce11'</span>,
    motif = <span class="st">'TT'</span>, 
    BPPARAM = <span class="fu"><a href="../reference/setUpBPPARAM.html">setUpBPPARAM</a></span>(<span class="fl">1</span>)
)
<span class="co">#&gt; - Mapping k-mers.</span>
<span class="co">#&gt; - 523903 pairwise distances measured.</span>
<span class="co">#&gt; - Calculating pairwise distance distribution.</span>
<span class="co">#&gt; - Normalizing distogram vector.</span>
<span class="co">#&gt; - Applying Fast Fourier Transform to the normalized distogram.</span>
</pre></div>
<p>The main output of <code><a href="../reference/getPeriodicity.html">getPeriodicity()</a></code> is a table of power spectral density (PSD) values associated with discrete frequencies, computed using a Fast Fourier Transform. For a given frequency, a high PSD score indicates a high periodicity of the k-mer of interest.</p>
<p>In the previous example, TT dinucleotides in sequences around TSSs of ubiquitous promoters are highly periodic, with a periodicity of 10 bp.</p>
<div class="sourceCode" id="cb2"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="kw">periodicity_result</span><span class="op">$</span><span class="kw">PSD</span>)
<span class="co">#&gt;    freq    period          PSD</span>
<span class="co">#&gt; 1 0.005 200.00000 6.256976e-08</span>
<span class="co">#&gt; 2 0.010 100.00000 2.204282e-08</span>
<span class="co">#&gt; 3 0.015  66.66667 2.215522e-09</span>
<span class="co">#&gt; 4 0.020  50.00000 1.108237e-08</span>
<span class="co">#&gt; 5 0.025  40.00000 4.649689e-09</span>
<span class="co">#&gt; 6 0.030  33.33333 2.661198e-08</span>
<span class="fu"><a href="https://rdrr.io/r/base/subset.html">subset</a></span>(<span class="kw">periodicity_result</span><span class="op">$</span><span class="kw">periodicityMetrics</span>, <span class="kw">Period</span> <span class="op">==</span> <span class="fl">10</span>)
<span class="co">#&gt;    Freq Period          PSD</span>
<span class="co">#&gt; 20  0.1     10 3.633071e-06</span>
</pre></div>
<p>Graphical output of <code><a href="../reference/getPeriodicity.html">getPeriodicity()</a></code> can be obtained using the <code><a href="../reference/plotPeriodicityResults.html">plotPeriodicityResults()</a></code> function:</p>
<div class="sourceCode" id="cb3"><pre class="downlit">
<span class="fu"><a href="../reference/plotPeriodicityResults.html">plotPeriodicityResults</a></span>(<span class="kw">periodicity_result</span>)
</pre></div>
<p><img src="periodicDNA_files/figure-html/unnamed-chunk-5-1.png" width="100%"></p>
<p>The first plot shows the raw distribution of distances between pairs of ‘TT’ in the sequences of the provided GRanges. The second plot shows the decay-normalised distribution. Finally, the third plot shows the PSD scores of the ‘TT’ k-mer, measured from the normalised distribution.</p>
</div>
<div id="repeated-shuffling-of-input-sequences" class="section level4">
<h4 class="hasAnchor">
<a href="#repeated-shuffling-of-input-sequences" class="anchor"></a>Repeated shuffling of input sequences</h4>
<p>periodicDNA provides an approach to compare the periodicity of a given k-mer in a set of sequences compared to background. For a given k-mer at a period T in a set of input sequences, the fold-change over background of its PSD is estimated by iteratively shuffling the input sequences and estimating the resulting PSD values.<br>
Eventually, the log2 fold-change (l2FC) between the observed PSD and the median of the PSD values measured after shuffling is computed as follows:</p>
<p>l2FC = log2(observed PSD / median(shuffled PSDs)).</p>
<div class="sourceCode" id="cb4"><pre class="downlit">
<span class="kw">periodicity_result</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getPeriodicity.html">getPeriodicity</a></span>(
    <span class="kw">ce11_TSSs</span>[[<span class="st">'Ubiq.'</span>]][<span class="fl">1</span><span class="op">:</span><span class="fl">500</span>],
    genome = <span class="st">'BSgenome.Celegans.UCSC.ce11'</span>,
    motif = <span class="st">'TT'</span>, 
    n_shuffling = <span class="fl">5</span>
)
<span class="co">#&gt; - Calculating observed PSD</span>
<span class="co">#&gt; - Mapping k-mers.</span>
<span class="co">#&gt; - 523903 pairwise distances measured.</span>
<span class="co">#&gt; - Calculating pairwise distance distribution.</span>
<span class="co">#&gt; - Normalizing distogram vector.</span>
<span class="co">#&gt; - Applying Fast Fourier Transform to the normalized distogram.</span>
<span class="co">#&gt; - Shuffling 1/5</span>
<span class="co">#&gt; - Shuffling 2/5</span>
<span class="co">#&gt; - Shuffling 3/5</span>
<span class="co">#&gt; - Shuffling 4/5</span>
<span class="co">#&gt; - Shuffling 5/5</span>
<span class="co">#&gt; Only 5 shufflings. Cannot compute accurate empirical p-values. To compute empirical p-values, set up n_shuffling to at least 100. Only l2FC values are returned</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="kw">periodicity_result</span><span class="op">$</span><span class="kw">periodicityMetrics</span>)
<span class="co">#&gt;    Freq    Period PSD_observed        l2FC pval fdr</span>
<span class="co">#&gt; 1 0.005 200.00000     6.26e-08 -1.37213353   NA  NA</span>
<span class="co">#&gt; 2 0.010 100.00000     2.20e-08 -1.12754062   NA  NA</span>
<span class="co">#&gt; 3 0.015  66.66667     2.22e-09 -2.47370908   NA  NA</span>
<span class="co">#&gt; 4 0.020  50.00000     1.11e-08  0.09044258   NA  NA</span>
<span class="co">#&gt; 5 0.025  40.00000     4.65e-09 -0.14473772   NA  NA</span>
<span class="co">#&gt; 6 0.030  33.33333     2.66e-08  0.19102196   NA  NA</span>
<span class="fu"><a href="https://rdrr.io/r/base/subset.html">subset</a></span>(<span class="kw">periodicity_result</span><span class="op">$</span><span class="kw">periodicityMetrics</span>, <span class="kw">Period</span> <span class="op">==</span> <span class="fl">10</span>)
<span class="co">#&gt;    Freq Period PSD_observed     l2FC pval fdr</span>
<span class="co">#&gt; 20  0.1     10     3.63e-06 8.593242   NA  NA</span>
<span class="fu"><a href="../reference/plotPeriodicityResults.html">plotPeriodicityResults</a></span>(<span class="kw">periodicity_result</span>)
</pre></div>
<p><img src="periodicDNA_files/figure-html/unnamed-chunk-6-1.png" width="100%"></p>
<p>If <code>n_shuffling &gt;= 100</code>, an associated empirical p-value is calculated as well (North et al 2002). This metric indicates, for each individual period T, whether the observed PSD is significantly greater than the PSD values measured after shuffling the input sequences. Note that empirical p-values are only an estimation of the real p-value. Notably, small p-values are systematically under-estimated (North et al 2002).</p>
</div>
<div id="note" class="section level4">
<h4 class="hasAnchor">
<a href="#note" class="anchor"></a>Note</h4>
<p><code><a href="../reference/getPeriodicity.html">getPeriodicity()</a></code> can also be ran directly on a set of sequences of interest as follows:</p>
<div class="sourceCode" id="cb5"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span>(<span class="kw">ce11_proms_seqs</span>)
<span class="kw">periodicity_result</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getPeriodicity.html">getPeriodicity</a></span>(
    <span class="kw">ce11_proms_seqs</span>,
    motif = <span class="st">'TT'</span>, 
    BPPARAM = <span class="fu"><a href="../reference/setUpBPPARAM.html">setUpBPPARAM</a></span>(<span class="fl">1</span>)
)
<span class="co">#&gt; - Mapping k-mers.</span>
<span class="co">#&gt; - 117630 pairwise distances measured.</span>
<span class="co">#&gt; - Calculating pairwise distance distribution.</span>
<span class="co">#&gt; - Normalizing distogram vector.</span>
<span class="co">#&gt; - Applying Fast Fourier Transform to the normalized distogram.</span>
<span class="fu"><a href="https://rdrr.io/r/base/subset.html">subset</a></span>(<span class="kw">periodicity_result</span><span class="op">$</span><span class="kw">periodicityMetrics</span>, <span class="kw">Period</span> <span class="op">==</span> <span class="fl">10</span>)
<span class="co">#&gt;    Freq Period         PSD</span>
<span class="co">#&gt; 20  0.1     10 1.16806e-06</span>
</pre></div>
</div>
</div>
<div id="track-of-periodicity-over-a-set-of-genomic-ranges" class="section level3">
<h3 class="hasAnchor">
<a href="#track-of-periodicity-over-a-set-of-genomic-ranges" class="anchor"></a>Track of periodicity over a set of Genomic Ranges</h3>
<p>The other aim of periodicDNA is to generate continuous linear tracks of k-mer periodicity strength over genomic loci of interest. <code><a href="../reference/getPeriodicityTrack.html">getPeriodicityTrack()</a></code> can be used to generate suck genomic tracks. In the following example,</p>
<div class="sourceCode" id="cb6"><pre class="downlit">
<span class="kw">WW_10bp_track</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getPeriodicityTrack.html">getPeriodicityTrack</a></span>(
    genome = <span class="st">'BSgenome.Celegans.UCSC.ce11'</span>,
    granges = <span class="kw">ce11_proms</span>, 
    motif = <span class="st">'WW'</span>,
    period = <span class="fl">10</span>,
    BPPARAM = <span class="fu"><a href="../reference/setUpBPPARAM.html">setUpBPPARAM</a></span>(<span class="fl">1</span>),
    bw_file = <span class="st">'WW-10-bp-periodicity_over-proms.bw'</span>
)
</pre></div>
<p>When plotted over sets of ubiquitous, germline or somatic TSSs, the resulting track clearly shows increase of WW 10-bp periodicity above the ubiquitous and germline TSSs, whereas somatic TSSs do not show such increase.</p>
<div class="sourceCode" id="cb7"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span>(<span class="kw">ce11_TSSs</span>)
<span class="fu"><a href="../reference/plotAggregateCoverage.html">plotAggregateCoverage</a></span>(
    <span class="kw">WW_10bp_track</span>, 
    <span class="kw">ce11_TSSs</span>, 
    xlab = <span class="st">'Distance from TSS'</span>,
    ylab = <span class="st">'10-bp periodicity strength (forward proms.)'</span>
)
</pre></div>
<p><img src="https://raw.githubusercontent.com/js2264/periodicDNA/master/man/figures/TT-10bp-periodicity_tissue-spe-TSSs.png" width="100%"></p>
</div>
<div id="session-info" class="section level2">
<h2 class="hasAnchor">
<a href="#session-info" class="anchor"></a>Session info</h2>
<div class="sourceCode" id="cb8"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo</a></span>()
<span class="co">#&gt; R version 4.0.2 (2020-06-22)</span>
<span class="co">#&gt; Platform: x86_64-pc-linux-gnu (64-bit)</span>
<span class="co">#&gt; Running under: Ubuntu 18.04.5 LTS</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Matrix products: default</span>
<span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/openblas/libblas.so.3</span>
<span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/libopenblasp-r0.2.20.so</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; locale:</span>
<span class="co">#&gt;  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span>
<span class="co">#&gt;  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span>
<span class="co">#&gt;  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span>
<span class="co">#&gt;  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span>
<span class="co">#&gt;  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span>
<span class="co">#&gt; [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; attached base packages:</span>
<span class="co">#&gt; [1] parallel  stats4    stats     graphics  grDevices utils     datasets </span>
<span class="co">#&gt; [8] methods   base     </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; other attached packages:</span>
<span class="co">#&gt;  [1] BSgenome.Celegans.UCSC.ce11_1.4.2 periodicDNA_0.99.17              </span>
<span class="co">#&gt;  [3] BiocParallel_1.23.2               BSgenome_1.57.6                  </span>
<span class="co">#&gt;  [5] rtracklayer_1.49.5                Biostrings_2.57.2                </span>
<span class="co">#&gt;  [7] XVector_0.29.3                    magrittr_1.5                     </span>
<span class="co">#&gt;  [9] ggplot2_3.3.2                     GenomicRanges_1.41.6             </span>
<span class="co">#&gt; [11] GenomeInfoDb_1.25.11              IRanges_2.23.10                  </span>
<span class="co">#&gt; [13] S4Vectors_0.27.12                 BiocGenerics_0.35.4              </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; loaded via a namespace (and not attached):</span>
<span class="co">#&gt;  [1] SummarizedExperiment_1.19.6 zoo_1.8-8                  </span>
<span class="co">#&gt;  [3] xfun_0.16                   lattice_0.20-41            </span>
<span class="co">#&gt;  [5] colorspace_1.4-1            vctrs_0.3.4                </span>
<span class="co">#&gt;  [7] htmltools_0.5.0             yaml_2.2.1                 </span>
<span class="co">#&gt;  [9] XML_3.99-0.5                rlang_0.4.7                </span>
<span class="co">#&gt; [11] pkgdown_1.6.0               pillar_1.4.6               </span>
<span class="co">#&gt; [13] glue_1.4.2                  withr_2.2.0                </span>
<span class="co">#&gt; [15] matrixStats_0.56.0          GenomeInfoDbData_1.2.3     </span>
<span class="co">#&gt; [17] lifecycle_0.2.0             stringr_1.4.0              </span>
<span class="co">#&gt; [19] zlibbioc_1.35.0             munsell_0.5.0              </span>
<span class="co">#&gt; [21] gtable_0.3.0                ragg_0.3.1                 </span>
<span class="co">#&gt; [23] memoise_1.1.0               evaluate_0.14              </span>
<span class="co">#&gt; [25] labeling_0.3                Biobase_2.49.1             </span>
<span class="co">#&gt; [27] knitr_1.29                  scales_1.1.1               </span>
<span class="co">#&gt; [29] backports_1.1.9             DelayedArray_0.15.7        </span>
<span class="co">#&gt; [31] desc_1.2.0                  cpp11_0.2.1                </span>
<span class="co">#&gt; [33] farver_2.0.3                systemfonts_0.3.1          </span>
<span class="co">#&gt; [35] fs_1.5.0                    Rsamtools_2.5.3            </span>
<span class="co">#&gt; [37] digest_0.6.25               stringi_1.4.6              </span>
<span class="co">#&gt; [39] cowplot_1.1.0               grid_4.0.2                 </span>
<span class="co">#&gt; [41] rprojroot_1.3-2             tools_4.0.2                </span>
<span class="co">#&gt; [43] bitops_1.0-6                RCurl_1.98-1.2             </span>
<span class="co">#&gt; [45] tibble_3.0.3                crayon_1.3.4               </span>
<span class="co">#&gt; [47] pkgconfig_2.0.3             ellipsis_0.3.1             </span>
<span class="co">#&gt; [49] Matrix_1.2-18               assertthat_0.2.1           </span>
<span class="co">#&gt; [51] rmarkdown_2.3               R6_2.4.1                   </span>
<span class="co">#&gt; [53] GenomicAlignments_1.25.3    compiler_4.0.2</span>
</pre></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by <a href="https://js2264.github.io/">Jacques Serizay</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.0.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
